android/src/main/java/com/tailscale/ipn/DnsConfig.java

```diff
@@ -44,14 +41,14 @@ private String getDnsConfigs() {
         }
     }
 
-    void updateDNSFromNetwork(String dnsConfigs) {
+    boolean updateDNSFromNetwork(String dnsConfigs) {
         synchronized (this) {
-            this.dnsConfigs = dnsConfigs;
+            if (!dnsConfigs.equals(this.dnsConfigs)) {
```

Comment:
for empty `dnsConfigs`, "localhost" on Android works; which usually forwards queries to either network-provided DNS (dhcp et al) or one of the pre-configured fallback DNSes as set by the OEM (usually `8.8.8.8`) on behalf of the VPN app.

Reply:
Thanks, I'll give that a try!

Reply:
Unfortunately, that did not work, at least not on a Samsung Galaxy S23.

Reply:
Surprising. In our case, the VPN engine's (here, `Libtailscale`) stub resolver must forward incoming DNS packets it gets to `localhost:53` when it doesn't know the identity of underlying DNS.